index.js

```diff
@@ -540,6 +540,17 @@ const menuFactory = (core, proc, win) => {
         onclick: () => emitter('filemanager:menu:download')
       }
     ];
+
+    const editMenuMiddlewares = core
```

Comment:
One final thing to ensure this works on old versions of the client library:

```
   const editMenuMiddlewares = core.has('osjs/middleware')
     ? core.make('osjs/middleware')
          .get('osjs/filemanager:contextmenu:edit')
     : [];
```

Reply:
This could maybe be split out into a generic function that takes an argument for the `.get()` that is done there for easier integration in other menus in the future.

Reply:
Do you mean to move the whole process of appending to a separate function or only the code you suggested [here](https://github.com/os-js/osjs-filemanager-application/pull/31#discussion_r559228286)?

Reply:
We can write a function (say `addMiddlewareItems`) and use it like this:

```js
const menuItems = {
    file: addMiddlewareItems(createFileMenu),
    edit: addMiddlewareItems(createEditMenu),
    view: addMiddlewareItems(createViewMenu),
    go: addMiddlewareItems(createGoMenu)
};
```

Reply:
Yeah, like this:

```
const menuFactory = (core, proc, win) => {

// ...
  const menuItemsFromMiddleware = async (type) => {
    if (core.has('osjs/middleware')) {
      const items = core.make('osjs/middleware').get(`osjs/filemanager:contextmenu:${type}`);
      const promises = items
        .map(fn => fn({file: item, isContextMenu}));

     const resolved = await Promise.all(promises);
     const result = resolved
        .filter(items => items instanceof Array)

      return [].concat(...result);
    }
   return [];
  }

// ...
    const appendItems = await menuItemsFromMiddleware('edit');
    return [...editMenuItems, ...appendItems];
```

Reply:
I did not take into account the argument in `fn()` call here. This could simply be an argument of the wrapping function:

```
  const menuItemsFromMiddleware = async (type, arg) => {
```

```
        .map(fn => fn(arg);

```

Reply:
I also changed the middleware group name so that it can be used for the other menu types. Does it sound good?

Reply:
I was about to suggest that :D Good stuff!